src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -195,11 +216,47 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
      * @return String: comma separated list with leading space in square brackets
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    String buildFieldDataString(NamedWriteable builder, Set<Index> successfulSearchShardIndices) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            fieldDataList.add(fieldName);
+            fieldDataList.add(getFieldType(fieldName, successfulSearchShardIndices));
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Set<Index> successfulSearchShardIndices) {
+        for (Index index : successfulSearchShardIndices) {
+            Map<String, String> indexMap = fieldTypeMap.get(index);
+            if (indexMap != null) {
+                String fieldType = indexMap.get(fieldName);
+                if (fieldType != null) {
+                    return fieldType;
+                }
+            }
+        }
+        Map<String, MappingMetadata> allIndicesMap;
+        try {
+            allIndicesMap = clusterService.state()
+                .metadata()
+                .findMappings(successfulSearchShardIndices.stream().map(Index::getName).toArray(String[]::new), input -> (str -> true));
+        } catch (IOException e) {
+            return null;
+        }
+        for (Index index : successfulSearchShardIndices) {
+            MappingMetadata mappingMetadata = allIndicesMap.get(index.getName());
+            if (mappingMetadata != null) {
+                @SuppressWarnings("unchecked")
+                Map<String, ?> propertiesAsMap = (Map<String, ?>) mappingMetadata.getSourceAsMap().get("properties");
```

Comment:
Can we preserve this `propertiesAsMap` for an index for the search request? Don't really want to create parse the same mappingMetadata for getting different fieldTypes within same index